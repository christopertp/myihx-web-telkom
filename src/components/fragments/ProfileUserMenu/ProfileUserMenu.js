import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import Button from '../../elements/Button';
import Modal from '../../elements/Modal';
import Switch from '../../elements/Switch';
import { clearStorages } from '../../../utils/storage';
import ModalConfirmation from '../../elements/ModalConfirmation';
import styles from './styles.scoped.css';

export default function ProfileUserMenu() {
  const [modal, setModal] = useState(false);
  const [language, setLanguage] = useState('ID');
  const [confirmation, setConfirmation] = useState(false);
  const renderList = (title, to) => {
    return (
      <Link to={to || '#'}>
        <span>{title}</span>
        <img alt="chevron red" src="/assets/ic_chevron_red.svg" />
      </Link>
    );
  };

  const handleAccept = () => {
    const chooseLanguage = language === 'ID' ? 'EN' : 'ID';
    setLanguage(chooseLanguage);
    setConfirmation(!confirmation);
  };
  return (
    <>
      <section className={styles.root}>
        <p>Account</p>
        {renderList('Edit Profile', '/profile/edit')}
        {renderList('Change Password', '/profile/change-password')}
        {renderList('Change Phone Number', '/profile/change-mobile')}
        {renderList('Change Email', '/profile/change-email')}
        {renderList('Manage Users', '/manage-users')}
        {renderList('Manage Indihome Numbers')}
        <p>Membership</p>
        {renderList('About Poin myIndiHome', '/profile/about-point')}
        {renderList('Point History')}
        {renderList('Merchants & Benefits')}
        {renderList('Browse Rewards', '/rewards/browse')}
        {renderList('My Vouchers')}
        <p>Other</p>
        {renderList('Manage T-Money')}
        {renderList('Manage Subscriptions', '/manage-subscriptions')}
        {renderList('Notification Settings', '/notification-settings')}
        <div className={styles['hide-desktop']}>
          <span>Switch Language</span>
          <Switch
            className={styles['switch']}
            inputProps={{
              id: 'language',
              checked: language === 'ID',
              onChange: () => {
                setConfirmation(!confirmation);
              },
            }}
            text={['ID', 'EN']}
          />
        </div>
        {renderList('Send Feedback', '/send-feedback')}
        {renderList('About the Application', '/about')}
        <Button onClick={() => setModal(true)}>logout</Button>
      </section>
      <ModalLogOut onClose={() => setModal(false)} open={modal} />
      <ModalConfirmation
        onAccept={handleAccept}
        onClose={() => setConfirmation(false)}
        onRefuse={() => setConfirmation(false)}
        open={confirmation}
        reverseBtn
        subtitle="This process may take a few seconds."
        title={`Switch language to ${language === 'ID' ? 'Bahasa Indonesia' : 'English'}`}
      />
    </>
  );
}
export function ModalLogOut(props) {
  const { onClose, open } = props;
  const onClick = () => {
    clearStorages();
    location.href = '/';
  };

  return (
    <Modal className={styles.modal} onClose={onClose} open={open}>
      <h3>Are you want to log out from myIndiHome?</h3>
      <p>You will need to login again next time you open myIndiHome app.</p>
      <div>
        <Button onClick={onClick} variant="bordered">
          yes
        </Button>
        <Button onClick={onClose}>no</Button>
      </div>
    </Modal>
  );
}

ModalLogOut.defaultProps = {
  onClose: () => {},
  open: false,
};

ModalLogOut.propTypes = {
  onClose: PropTypes.func,
  open: PropTypes.bool,
};
