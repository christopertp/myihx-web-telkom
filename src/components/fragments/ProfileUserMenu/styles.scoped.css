.root {
  background-color: #ffffff;
  border-radius: 0.5rem;
  box-shadow: 0 0.5rem 1rem rgba(0, 0, 1, 0.15);
  margin-bottom: 4rem;
  padding: 1.5rem;
}

.root > p {
  color: #999999;
  font-size: 0.9rem;
  letter-spacing: 0.0625rem;
  margin: 1rem 0;
  text-transform: uppercase;
}

.root > p:first-of-type {
  margin-top: 0;
}

.root > a {
  align-items: center;
  display: flex;
  justify-content: space-between;
  padding: 1.25rem 0;
}

.root > a:hover > span {
  color: var(--primary-color);
}

.root > a > span,
.root > div > span {
  color: #333435;
  font-size: 1.125rem;
  font-weight: 800;
  transition: color 0.15s ease-in;
}

.root > a > img {
  height: 1.3rem;
}

.root > div {
  align-items: center;
  display: flex;
  justify-content: space-between;
  padding: 1.25rem 0;
}

.root > button {
  margin-top: 2rem;
  width: 100%;
}

.modal {
  text-align: center;
}

.modal > h3 {
  font-size: 1.75rem;
  line-height: 1.2;
  margin-bottom: 0.75rem;
}

.modal > p {
  color: #808080;
  line-height: 1.6;
  margin-bottom: 1.5rem;
}

.modal > div {
  display: flex;
  justify-content: space-between;
  text-transform: uppercase;
}

.hide-desktop {
  display: none !important;
}

.modal > div > button {
  flex: 1;
}

.modal > div > button:first-child {
  margin-right: 1.25rem;
}

.switch {
  background-color: #115b79;
}
.switch:not(:checked) {
  background-color: #115b79;
}

@media only screen and (max-width: 48em) {
  .root {
    margin: 0 -1rem -1rem;
    padding: 1rem;
  }
  .root > p {
    font-size: 0.8rem;
    margin-bottom: 0.75rem;
  }
  .root > a {
    border-bottom: 0.063rem solid #ededed;
    padding: 0.75rem 0.5rem 0.75rem 0;
  }
  .root > a > span,
  .root > div > span {
    font-size: 1rem;
  }
  .root > a > img {
    height: 1.2rem;
  }

  .root > div {
    border-bottom: 0.063rem solid #ededed;
    padding: 0.75rem 0.5rem 0.75rem 0;
  }
  .modal > h3 {
    font-size: 1.25rem;
  }
  .modal > p {
    font-size: 0.8rem;
  }
  .hide-desktop {
    display: flex !important;
  }
}
