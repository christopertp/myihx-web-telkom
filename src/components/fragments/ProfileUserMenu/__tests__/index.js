import React, { useState } from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import ProfileUserMenu, { ModalLogOut } from '../ProfileUserMenu';

describe('src/components/fragments/ProfileUserMenu', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<ProfileUserMenu />);
    expect(tree).toMatchSnapshot();

    const setModal = jest.fn();
    useState.mockImplementationOnce(() => [false, setModal]);

    const result = ProfileUserMenu({
      ...ProfileUserMenu.defaultProps,
    });
    result.props.children[0].props.children[20].props.onClick();
    const result2 = ProfileUserMenu({
      ...ProfileUserMenu.defaultProps,
    });
    result2.props.children[1].props.onClose();
    expect(setModal).toHaveBeenCalled();
  });

  test('ModalLogout', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<ModalLogOut />);
    expect(tree).toMatchSnapshot();

    ModalLogOut.defaultProps.onClose();
    const result = ModalLogOut({
      ...ModalLogOut.defaultProps,
    });
    result.props.children[2].props.children[0].props.onClick();
    expect(result.props.children[2].props.children[0].props.variant).toBe('bordered');
  });

  test('Change language to indonesia', () => {
    const setLanguage = jest.fn();
    const setConfirmation = jest.fn();
    const setModal = jest.fn();

    useState
      .mockImplementationOnce((v) => [v, setModal])
      .mockImplementationOnce(() => ['ID', setLanguage])
      .mockImplementationOnce((v) => [v, setConfirmation]);

    const res = ProfileUserMenu();
    const { children } = res.props;

    children[0].props.children[17].props.children[1].props.inputProps.onChange();
    expect(setConfirmation).toHaveBeenNthCalledWith(1, true);

    children[2].props.onAccept();
    expect(setLanguage).toHaveBeenNthCalledWith(1, 'EN');
    expect(setConfirmation).toHaveBeenNthCalledWith(2, true);

    expect(children[2].props.title).toBe('Switch language to Bahasa Indonesia');

    children[2].props.onClose();
    expect(setConfirmation).toHaveBeenNthCalledWith(3, false);

    children[2].props.onRefuse();
    expect(setConfirmation).toHaveBeenNthCalledWith(4, false);
  });

  test('Change language to english', () => {
    const setLanguage = jest.fn();
    const setConfirmation = jest.fn();
    const setModal = jest.fn();

    useState
      .mockImplementationOnce((v) => [v, setModal])
      .mockImplementationOnce(() => ['EN', setLanguage])
      .mockImplementationOnce((v) => [v, setConfirmation]);

    const res = ProfileUserMenu();
    const { children } = res.props;

    children[0].props.children[17].props.children[1].props.inputProps.onChange();
    expect(setConfirmation).toHaveBeenNthCalledWith(1, true);

    children[2].props.onAccept();
    expect(setLanguage).toHaveBeenNthCalledWith(1, 'ID');
    expect(setConfirmation).toHaveBeenNthCalledWith(2, true);

    expect(children[2].props.title).toBe('Switch language to English');
  });
});
