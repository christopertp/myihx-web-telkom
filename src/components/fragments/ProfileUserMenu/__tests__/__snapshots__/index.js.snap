// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/fragments/ProfileUserMenu ModalLogout 1`] = `
<Modal
  className=""
  onClose={[Function]}
  open={false}
>
  <h3>
    Are you want to log out from myIndiHome?
  </h3>
  <p>
    You will need to login again next time you open myIndiHome app.
  </p>
  <div>
    <Button
      className=""
      disabled={false}
      onClick={[Function]}
      to=""
      type="button"
      variant="bordered"
    >
      yes
    </Button>
    <Button
      className=""
      disabled={false}
      onClick={[Function]}
      to=""
      type="button"
      variant="red"
    >
      no
    </Button>
  </div>
</Modal>
`;

exports[`src/components/fragments/ProfileUserMenu render 1`] = `
<React.Fragment>
  <section>
    <p>
      Account
    </p>
    <Link
      to="/profile/edit"
    >
      <span>
        Edit Profile
      </span>
      <img
        alt="chevron red"
        src="/assets/ic_chevron_red.svg"
      />
    </Link>
    <Link
      to="/profile/change-password"
    >
      <span>
        Change Password
      </span>
      <img
        alt="chevron red"
        src="/assets/ic_chevron_red.svg"
      />
    </Link>
    <Link
      to="/profile/change-mobile"
    >
      <span>
        Change Phone Number
      </span>
      <img
        alt="chevron red"
        src="/assets/ic_chevron_red.svg"
      />
    </Link>
    <Link
      to="/profile/change-email"
    >
      <span>
        Change Email
      </span>
      <img
        alt="chevron red"
        src="/assets/ic_chevron_red.svg"
      />
    </Link>
    <Link
      to="/manage-users"
    >
      <span>
        Manage Users
      </span>
      <img
        alt="chevron red"
        src="/assets/ic_chevron_red.svg"
      />
    </Link>
    <Link
      to="#"
    >
      <span>
        Manage Indihome Numbers
      </span>
      <img
        alt="chevron red"
        src="/assets/ic_chevron_red.svg"
      />
    </Link>
    <p>
      Membership
    </p>
    <Link
      to="/profile/about-point"
    >
      <span>
        About Poin myIndiHome
      </span>
      <img
        alt="chevron red"
        src="/assets/ic_chevron_red.svg"
      />
    </Link>
    <Link
      to="#"
    >
      <span>
        Point History
      </span>
      <img
        alt="chevron red"
        src="/assets/ic_chevron_red.svg"
      />
    </Link>
    <Link
      to="#"
    >
      <span>
        Merchants & Benefits
      </span>
      <img
        alt="chevron red"
        src="/assets/ic_chevron_red.svg"
      />
    </Link>
    <Link
      to="/rewards/browse"
    >
      <span>
        Browse Rewards
      </span>
      <img
        alt="chevron red"
        src="/assets/ic_chevron_red.svg"
      />
    </Link>
    <Link
      to="#"
    >
      <span>
        My Vouchers
      </span>
      <img
        alt="chevron red"
        src="/assets/ic_chevron_red.svg"
      />
    </Link>
    <p>
      Other
    </p>
    <Link
      to="#"
    >
      <span>
        Manage T-Money
      </span>
      <img
        alt="chevron red"
        src="/assets/ic_chevron_red.svg"
      />
    </Link>
    <Link
      to="/manage-subscriptions"
    >
      <span>
        Manage Subscriptions
      </span>
      <img
        alt="chevron red"
        src="/assets/ic_chevron_red.svg"
      />
    </Link>
    <Link
      to="/notification-settings"
    >
      <span>
        Notification Settings
      </span>
      <img
        alt="chevron red"
        src="/assets/ic_chevron_red.svg"
      />
    </Link>
    <div>
      <span>
        Switch Language
      </span>
      <Switch
        className=""
        inputProps={
          Object {
            "checked": true,
            "id": "language",
            "onChange": [Function],
          }
        }
        text={
          Array [
            "ID",
            "EN",
          ]
        }
      />
    </div>
    <Link
      to="/send-feedback"
    >
      <span>
        Send Feedback
      </span>
      <img
        alt="chevron red"
        src="/assets/ic_chevron_red.svg"
      />
    </Link>
    <Link
      to="/about"
    >
      <span>
        About the Application
      </span>
      <img
        alt="chevron red"
        src="/assets/ic_chevron_red.svg"
      />
    </Link>
    <Button
      className=""
      disabled={false}
      onClick={[Function]}
      to=""
      type="button"
      variant="red"
    >
      logout
    </Button>
  </section>
  <ModalLogOut
    onClose={[Function]}
    open={false}
  />
  <ModalConfirmation
    onAccept={[Function]}
    onClose={[Function]}
    onRefuse={[Function]}
    open={false}
    reverseBtn={true}
    subtitle="This process may take a few seconds."
    title="Switch language to Bahasa Indonesia"
  />
</React.Fragment>
`;
